### PR TITLE
Req resp bitswap

### LINKED ISSUE #4: implement optimizations from netflix
https://blog.ipfs.io/2020-02-14-improved-bitswap-for-container-distribution/

### LINKED ISSUE #5: make sure the maximum block size is handled correctly

### LINKED ISSUE #6: Panic in Bitswap::ledger() unwrap
I just saw this unwrap() panic:

```
     61     fn ledger(&mut self, peer_id: &PeerId) -> &mut Ledger<MH> {
     62         self.connected_peers.get_mut(peer_id).unwrap()
     63     }
```

Not exactly sure what happened yet but peers may be disconnecting unexpectedly because I'm hitting the file descriptor limit. In any case, this unwrap seems like a problem.

### LINKED ISSUE #7: limit to one bitswap substream per connection

### PR DESCRIPTION
**Pull request by dvc94ch**
Closes #4 closes #5 close #6 close #7

TODO:
- [x] support canceling queries
- [x] box bitswap store

### CODE REVIEW

File: src/behaviour.rs [hunk 1]
```diff
@@ -5,263 +5,368 @@
...
-    target_peers: FnvHashSet<PeerId>,
-    /// Ledger
-    connected_peers: HashMap<PeerId, Ledger<MH>>,
-    /// Wanted blocks
-    wanted_blocks: HashMap<Cid, Priority>,
+/// Bitswap configuration.
+pub struct BitswapConfig<P, S> {
+    _marker: PhantomData<P>,
+    store: S,
+    pub credit_timeout: Duration,
```
Comment (dvc94ch) [inline src/behaviour.rs:48]: Unused, remove

**Comment by rklaehn** [comment]
If you want to get good reviews, it might be useful to add a module level documentation about what the purpose of the thing is. Especially the query stuff. The normal bitswap have/want stuff is pretty obvious.

**Comment by dvc94ch** [comment]
Added some comments on future work here #10, but I think it's good enough for the time being. Also added some docs, but really they're two rather simple finite state machines (one per query type) and a query manager that manages them.

**Comment by dvc94ch** [comment]
Good enough for now, works with ipfs-embed aliasing.

**Comment by rklaehn** [comment]
Thanks for the comment. Looks good. Agree about the boxing of the BitswapStore.

**Pull request merged**